src/com/google/enterprise/adaptor/database/ResponseGenerator.java

```diff
@@ -386,20 +402,121 @@ public void generateResponse(ResultSet rs, Response resp)
+
       OutputStream out = resp.getOutputStream();
-      com.google.enterprise.adaptor.IOHelper.copyStream(in, out);
-      in.close();
-      if (!(blob instanceof javax.sql.rowset.serial.SerialBlob)) {
-        // SerialBlob is adamant about not supporting free
-        try {
-          blob.free();
-        } catch (java.sql.SQLFeatureNotSupportedException | 
-            UnsupportedOperationException unsupported) {
-          // let JVM garbage collection deal with it
-        }
+      switch (columnType) {
+        case Types.BLOB:
+          Blob blob = rs.getBlob(index);
+          if (blob != null) {
+            try (InputStream in = blob.getBinaryStream()) {
+              IOHelper.copyStream(in, out);
+            } finally {
+              try {
+                blob.free();
+              } catch (Exception e) {
+                log.log(Level.FINEST, "Error closing BLOB", e);
+              }
+            }
+          }
+          break;
+        case Types.CLOB:
+          Clob clob = rs.getClob(index);
+          if (clob != null) {
+            try (Reader reader = clob.getCharacterStream();
+                Writer writer = new OutputStreamWriter(out, UTF_8)) {
+              copy(reader, writer);
+            } finally {
+              try {
+                clob.free();
+              } catch (Exception e) {
+                log.log(Level.FINEST, "Error closing CLOB", e);
+              }
+            }
+          }
+          break;
+        case Types.NCLOB:
+          NClob nclob = rs.getNClob(index);
+          if (nclob != null) {
+            try (Reader reader = nclob.getCharacterStream();
+                Writer writer = new OutputStreamWriter(out, UTF_8)) {
+                copy(reader, writer);
+            } finally {
+              try {
+                nclob.free();
+              } catch (Exception e) {
+                log.log(Level.FINEST, "Error closing NCLOB", e);
+              }
+            }
+          }
+          break;
+        case Types.LONGVARBINARY:
+          try (InputStream in = rs.getBinaryStream(index)) {
+            if (in != null) {
+              IOHelper.copyStream(in, out);
+            }
+          }
+          break;
+        case Types.LONGVARCHAR:
+          try (Reader reader = rs.getCharacterStream(index);
+              Writer writer = new OutputStreamWriter(out, UTF_8)) {
+            if (reader != null) {
+              copy(reader, writer);
+            }
+          }
+          break;
+        case Types.LONGNVARCHAR:
+          try (Reader reader = rs.getNCharacterStream(index);
+              Writer writer = new OutputStreamWriter(out, UTF_8)) {
+            if (reader != null) {
+              copy(reader, writer);
+            }
+          }
+          break;
+        case Types.BINARY:
+        case Types.VARBINARY:
+          byte[] b = rs.getBytes(index);
+          if (b != null) {
+            out.write(b);
+          }
+          break;
+        case Types.TINYINT:
```

Comment:
As discussed, let's drop the integer types, at least for now. For posterity, we're also not handling dates and structured types here. The content must be of a binary or character type.